packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -0,0 +1,438 @@
+import { parseVatSlot } from './parseVatSlots.js';
+
+/*
+  Theory of Operation (imports/Presences)
+
+    Imports are in one of 5 states: UNKNOWN, REACHABLE, UNREACHABLE,
+    COLLECTED, FINALIZED. Note that there's no actual state machine with those
+    values, and we can't observe all of the transitions from JavaScript, but
+    we can describe what operations could cause a transition, and what our
+    observations allow us to deduce about the state:
+
+    * UNKNOWN moves to REACHABLE when a crank introduces a new import
+    * userspace holds a reference only in REACHABLE
+    * REACHABLE moves to UNREACHABLE only during a userspace crank
+    * UNREACHABLE moves to COLLECTED when GC runs, which queues the finalizer
+    * COLLECTED moves to FINALIZED when a new turn runs the finalizer
+    * liveslots moves from FINALIZED to UNKNOWN by syscalling dropImports
+
+    convertSlotToVal either imports a vref for the first time, or
+    re-introduces a previously-seen vref. It transitions from:
+
+    * UNKNOWN to REACHABLE by creating a new Presence
+    * UNREACHABLE to REACHABLE by re-using the old Presence that userspace
+      forgot about
+    * COLLECTED/FINALIZED to REACHABLE by creating a new Presence
+
+    Our tracking tables hold data that depends on the current state:
+
+    * slotToVal holds a WeakRef in [REACHABLE, UNREACHABLE, COLLECTED]
+    * that WeakRef .deref()s into something in [REACHABLE, UNREACHABLE]
+    * possiblyDeadSet holds the vref only in FINALIZED
+    * (TODO) re-introduction could remove the vref from possiblyDeadSet
+
+    Each state thus has a set of perhaps-measurable properties:
+
+    * UNKNOWN: slotToVal[baseRef] is missing, baseRef not in possiblyDeadSet
+    * REACHABLE: slotToVal has live weakref, userspace can reach
+    * UNREACHABLE: slotToVal has live weakref, userspace cannot reach
+    * COLLECTED: slotToVal[baseRef] has dead weakref
+    * FINALIZED: slotToVal[baseRef] is missing, baseRef is in possiblyDeadSet
+
+    Our finalizer callback is queued by the engine's transition from
+    UNREACHABLE to COLLECTED, but the baseRef might be re-introduced before the
+    callback has a chance to run. There might even be multiple copies of the
+    finalizer callback queued. So the callback must deduce the current state
+    and only perform cleanup (i.e. delete the slotToVal entry and add the
+    baseRef to possiblyDeadSet) in the COLLECTED state.
+
+    Our general rule is "trust the finalizer". The GC code below
+    considers a Presence to be reachable (the vref's "RAM pillar"
+    remains "up") until it moves to the FINALIZED state. We do this to
+    avoid race conditions between some other pillar dropping (and a
+    BOYD sampling the WeakRef) while it is in the COLLECTED state. If
+    we treated COLLECTED as unreachable,, then the subsequent
```

Comment:
Typo fix:
```suggestion
    we treated COLLECTED as unreachable, then the subsequent
```
Also, what is intended by the lowercase "unreachable" here? Rather than talking about states, should this say something like "_If we examined the vref in COLLECTED", then…"_?

Reply:
Lowercase because I got fuzzy with the states I'm trying to describe. The original GC terminology was established before we had virtual objects, so it was almost entirely talking about Presences. We have one set of states for the Presence object (where COLLECTED and FINALIZED are things), but we have a different set of states for the abstract vref-centric object thing. Within a vat, the vref thing might be reachable (by any one of the three pillars). From the kernel's point of view, the vat has a reachable or merely-recognizable or neither c-list import status.

Also, the phrase "if we treated COLLECTED as UNREACHABLE" seems non-sensical the previous several paragraphs describe these as mutually-exclusive states (along with REACHABLE and UNKNOWN and FINALIZED). What we're really saying is "if we treated the Presence `Object` having the engine-GC-category COLLECTED state as meaning the RAM pillar was *down* (and thus not contributing to the reachability of the vref-thing), then the subsequent finalizer ..".

I'll see if I can tweak it a bit without going that verbose. Some day we'll need a more precise writeup.

Reply:
Oh yeah I hint at that at the beginning of the paragraph:

> The GC code below considers a Presence to be reachable (the vref's "RAM pillar" remains "up") until it moves to the FINALIZED state.

Reply:
maybe:

> If we treated COLLECTED as unreachable (the RAM pillar being "down"), then the subsequent finalizer callback would..

Reply:
Yeah, that's good.